addons/website/models/website.py

```diff
@@ -1342,3 +1342,93 @@ def _get_cached_values(self):
 
     def _get_cached(self, field):
         return self._get_cached_values()[field]
+
+    def _get_html_fields(self):
+        html_fields = [('ir_ui_view', 'arch_db')]
+        cr = self.env.cr
+        cr.execute(r"""
+            SELECT f.model, f.name
+                FROM ir_model_fields f
+                JOIN ir_model m ON m.id = f.model_id
+            WHERE f.ttype = 'html'
+                AND f.store = true
+                AND m.transient = false
+                AND f.model NOT LIKE 'ir.actions%'
+                AND f.model != 'mail.message'
+        """)
+        for model, name in cr.fetchall():
+            table = self.env[model]._table
+            if tools.table_exists(cr, table) and tools.column_exists(cr, table, name):
+                html_fields.append((table, name))
+        return html_fields
+
+    def _get_snippets_assets(self):
+        """Returns every parent snippet asset from the database
+        in the form of a list of tuples :
+        [(snippet_module, snippet_id, asset_version, asset_type, asset_id)]
+        """
+        self.env.cr.execute(r"""
+            SELECT DISTINCT ON (snippet_id,
+                        asset_version,
+                        asset_type) regexp_matches[1] AS snippet_module,
+                    regexp_matches[2] AS snippet_id,
+                    regexp_matches[3] AS asset_version,
+                    CASE
+                        WHEN regexp_matches[4]='scss' THEN 'css'
+                        ELSE regexp_matches[4]
+                    END AS asset_type,
+                    id AS asset_id
+            FROM
+            (SELECT REGEXP_MATCHES(PATH, '(\w*)\/.*\/snippets\/(\w*)\/(\d{3})\.(js|scss)'),
+                    id
+            FROM ir_asset) AS regexp
+            ORDER BY snippet_id,
+                    asset_version,
+                    asset_type,
+                    asset_id;
```

Comment:
That indent :grimacing: 
I was trying to see where was the order by ID neeeded to make the DISTINCT work as we expect (keep only real template asset, not overide)
What about something like
```SQL
    SELECT DISTINCT ON (snippet_id, asset_version, asset_type)
           regexp_matches[1] AS snippet_module,
           regexp_matches[2] AS snippet_id,
           regexp_matches[3] AS asset_version,
           CASE
               WHEN regexp_matches[4]='scss' THEN 'css'
               ELSE regexp_matches[4]
           END AS asset_type,
           id AS asset_id
      FROM (
          SELECT REGEXP_MATCHES(PATH, '(\w*)\/.*\/snippets\/(\w*)\/(\d{3})\.(js|scss)'),
                 id
            FROM ir_asset
      ) AS regexp
     ORDER BY snippet_id, asset_version, asset_type, asset_id;
     ```
 Also, can you add a comment that the DISTINCT + the order by ID are there to filter/keep only original asset and filter out duplicate?
 
 Or it will be hard/make the next developer have to think a lot more than if you write the comment ;-)

Reply:
Noted, I'll stop using https://sqlformat.org/ which does not seem to make a lot of sense and stick to the odoo upgrade guidelines, sorry about that !